File: src/navidrome.ts
/**
 * Subsonic client for Navidrome — resolve album/track IDs, scan, build playlists.
 * Auth / envelope-checking / matching are pure functions (unit-tested without
 * network); the `Subsonic` class is the thin fetch wrapper around them.
 */

import { createHash, randomBytes } from "node:crypto";
import { existsSync, readFileSync } from "node:fs";

export const DEFAULT_CLIENT = "subwave-docs";
export const API_VERSION = "1.16.1";

export class SubsonicError extends Error {}

export interface Album {
  id: string;
  name: string;
  artist?: string;
  songCount?: number;
  song?: Song | Song[];
  [k: string]: unknown;
}

export interface Song {
  id: string;
  title: string;
  album?: string;
  artist?: string;
  track?: number;
  [k: string]: unknown;
}

// --- pure helpers ------------------------------------------------------------

export function subsonicToken(password: string, salt: string): string {
  return createHash("md5").update(password + salt).digest("hex");
}

export function authParams(user: string, password: string, salt: string,
                           client: string = DEFAULT_CLIENT,
                           version: string = API_VERSION): Record<string, string> {
  return { u: user, t: subsonicToken(password, salt), s: salt, v: version, c: client, f: "json" };
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function checkResponse(payload: any): any {
  const resp = payload?.["subsonic-response"];
  if (resp === undefined) {
    throw new SubsonicError("malformed response: no 'subsonic-response' envelope");
  }
  if (resp.status === "failed") {
    const err = resp.error ?? {};
    throw new SubsonicError(`Subsonic error ${err.code}: ${err.message}`);
  }
  return resp;
}

export function asList<T>(x: T | T[] | undefined | null): T[] {
  if (x === undefined || x === null) return [];
  return Array.isArray(x) ? x : [x];
}

export function matchAlbum(albums: Album[], album: string, artist?: string): Album | null {
  for (const a of albums) {
    if ((a.name ?? "").toLowerCase() === album.toLowerCase()
        && (artist === undefined || (a.artist ?? "").toLowerCase() === artist.toLowerCase())) {
      return a;
    }
  }
  return null;
}

export function songsOfAlbum(albumObj: { song?: Song | Song[] }): Song[] {
  return asList(albumObj.song);
}

export function matchSong(songs: Song[], title: string,
                          album?: string, artist?: string): Song | null {
  for (const s of songs) {
    if ((s.title ?? "").toLowerCase() === title.toLowerCase()
        && (album === undefined || (s.album ?? "").toLowerCase() === album.toLowerCase())
        && (artist === undefined || (s.artist ?? "").toLowerCase() === artist.toLowerCase())) {
      return s;
    }
  }
  return null;
}

/** Minimal .env loader (avoids a dotenv dependency). Existing env wins. */
export function loadDotenv(path: string): void {
  if (!existsSync(path)) return;
  for (const raw of readFileSync(path, "utf-8").split("\n")) {
    const line = raw.trim();
    if (!line || line.startsWith("#") || !line.includes("=")) continue;
    const i = line.indexOf("=");
    const key = line.slice(0, i).trim();
    if (process.env[key] === undefined) process.env[key] = line.slice(i + 1).trim();
  }
}

// --- client ------------------------------------------------------------------

export interface SubsonicConfig {
  baseUrl: string;
  user: string;
  password: string;
  client?: string;
  version?: string;
  timeoutMs?: number;
}

export class Subsonic {
  constructor(private readonly cfg: SubsonicConfig) {}

  private async request(method: string,
                        params: Record<string, string | string[]> = {}): Promise<any> {
    const salt = randomBytes(8).toString("hex");
    const auth = authParams(this.cfg.user, this.cfg.password, salt, this.cfg.client, this.cfg.version);
    const qs = new URLSearchParams();
    for (const [k, v] of Object.entries({ ...auth, ...params })) {
      if (Array.isArray(v)) v.forEach((x) => qs.append(k, x));
      else qs.append(k, v);
    }
    const url = `${this.cfg.baseUrl.replace(/\/+$/, "")}/rest/${method}.view?${qs.toString()}`;
    let payload: unknown;
    try {
      const res = await fetch(url, { signal: AbortSignal.timeout(this.cfg.timeoutMs ?? 15000) });
      payload = await res.json();
    } catch (e) {
      throw new SubsonicError(`HTTP error calling ${method}: ${String(e)}`);
    }
    return checkResponse(payload);
  }

  // read
  ping(): Promise<any> {
    return this.request("ping");
  }

  async search3(query: string, albumCount = 20, songCount = 30, artistCount = 0): Promise<any> {
    const r = await this.request("search3", {
      query,
      albumCount: String(albumCount),
      songCount: String(songCount),
      artistCount: String(artistCount),
    });
    return r.searchResult3 ?? {};
  }

  async findAlbum(album: string, artist?: string): Promise<Album | null> {
    const res = await this.search3(album, 50);
    return matchAlbum(asList(res.album), album, artist);
  }

  async getAlbum(id: string): Promise<Album> {
    return (await this.request("getAlbum", { id })).album ?? {};
  }

  async findSong(title: string, album?: string, artist?: string): Promise<Song | null> {
    const res = await this.search3(title, 20, 50);
    return matchSong(asList(res.song), title, album, artist);
  }

  async scanStatus(): Promise<any> {
    return (await this.request("getScanStatus")).scanStatus ?? {};
  }

  async getPlaylists(): Promise<any[]> {
    return asList((await this.request("getPlaylists")).playlists?.playlist);
  }

  // write (used only by the prompted playlist-build step)
  async startScan(full = false): Promise<any> {
    return (await this.request("startScan", full ? { fullScan: "true" } : {})).scanStatus ?? {};
  }

  async createPlaylist(name: string, songIds: string[]): Promise<any> {
    // Navidrome preserves the submitted songId order.
    return (await this.request("createPlaylist", { name, songId: songIds })).playlist ?? {};
  }
}

export function clientFromEnv(dotenvPath?: string): Subsonic {
  if (dotenvPath) loadDotenv(dotenvPath);
  const baseUrl = process.env.NAVIDROME_URL;
  const user = process.env.NAVIDROME_USER;
  const password = process.env.NAVIDROME_PASS;
  if (!baseUrl || !user || !password) {
    throw new SubsonicError("NAVIDROME_URL / NAVIDROME_USER / NAVIDROME_PASS not set (see .env.example)");
  }
  return new Subsonic({ baseUrl, user, password });
}

File: pipeline/navidrome.py
"""Subsonic client for Navidrome — resolve album/track IDs, scan, build playlists.

Covers the publish half of the pipeline (flow step 7b). Uses only the stdlib
(urllib), so it needs no venv. The auth, envelope-checking, and matching logic
are split out as pure functions so they unit-test without touching the network;
``Subsonic`` is the thin HTTP wrapper around them.

Credentials come from the environment (NAVIDROME_URL/USER/PASS; see .env.example).
Read-only in normal use here — the shared `david` account must never be rotated
from this repo (it fans out to SUB/WAVE + Homepage; see the vault note).
"""

from __future__ import annotations

import hashlib
import json
import os
import secrets
import urllib.error
import urllib.parse
import urllib.request
from dataclasses import dataclass
from pathlib import Path

DEFAULT_CLIENT = "subwave-docs"
API_VERSION = "1.16.1"


class SubsonicError(Exception):
    pass


# --- pure helpers (no network; directly unit-tested) -------------------------

def subsonic_token(password: str, salt: str) -> str:
    """Subsonic token auth: md5(password + salt)."""
    return hashlib.md5((password + salt).encode("utf-8")).hexdigest()


def auth_params(user: str, password: str, salt: str,
                client: str = DEFAULT_CLIENT, version: str = API_VERSION) -> dict:
    return {
        "u": user,
        "t": subsonic_token(password, salt),
        "s": salt,
        "v": version,
        "c": client,
        "f": "json",
    }


def check_response(payload: dict) -> dict:
    """Unwrap the ``subsonic-response`` envelope, raising on a failed status."""
    resp = payload.get("subsonic-response")
    if resp is None:
        raise SubsonicError("malformed response: no 'subsonic-response' envelope")
    if resp.get("status") == "failed":
        err = resp.get("error", {})
        raise SubsonicError(f"Subsonic error {err.get('code')}: {err.get('message')}")
    return resp


def as_list(x) -> list:
    """Subsonic JSON collapses single-element arrays to a bare object."""
    if x is None:
        return []
    return x if isinstance(x, list) else [x]


def match_album(albums: list, album: str, artist: str | None = None) -> dict | None:
    for a in albums:
        if a.get("name", "").lower() == album.lower() and (
            artist is None or a.get("artist", "").lower() == artist.lower()
        ):
            return a
    return None


def songs_of_album(album_obj: dict) -> list:
    return as_list(album_obj.get("song"))


def match_song(songs: list, title: str,
               album: str | None = None, artist: str | None = None) -> dict | None:
    for s in songs:
        if (s.get("title", "").lower() == title.lower()
                and (album is None or s.get("album", "").lower() == album.lower())
                and (artist is None or s.get("artist", "").lower() == artist.lower())):
            return s
    return None


# --- minimal .env loader (avoids a python-dotenv dependency) -----------------

def load_dotenv(path: str | Path) -> None:
    p = Path(path)
    if not p.exists():
        return
    for line in p.read_text(encoding="utf-8").splitlines():
        line = line.strip()
        if not line or line.startswith("#") or "=" not in line:
            continue
        key, val = line.split("=", 1)
        os.environ.setdefault(key.strip(), val.strip())


# --- client ------------------------------------------------------------------

@dataclass
class Subsonic:
    base_url: str
    user: str
    password: str
    client: str = DEFAULT_CLIENT
    version: str = API_VERSION
    timeout: int = 15

    def _request(self, method: str, **params) -> dict:
        salt = secrets.token_hex(8)
        query = {**auth_params(self.user, self.password, salt, self.client, self.version),
                 **params}
        url = (f"{self.base_url.rstrip('/')}/rest/{method}.view?"
               + urllib.parse.urlencode(query, doseq=True))
        try:
            with urllib.request.urlopen(url, timeout=self.timeout) as r:
                payload = json.loads(r.read().decode("utf-8"))
        except urllib.error.URLError as e:
            raise SubsonicError(f"HTTP error calling {method}: {e}") from e
        return check_response(payload)

    # read
    def ping(self) -> dict:
        return self._request("ping")

    def search3(self, query: str, album_count: int = 20,
                song_count: int = 30, artist_count: int = 0) -> dict:
        r = self._request("search3", query=query, albumCount=album_count,
                          songCount=song_count, artistCount=artist_count)
        return r.get("searchResult3", {})

    def find_album(self, album: str, artist: str | None = None) -> dict | None:
        res = self.search3(album, album_count=50)
        return match_album(as_list(res.get("album")), album, artist)

    def get_album(self, album_id: str) -> dict:
        return self._request("getAlbum", id=album_id).get("album", {})

    def find_song(self, title: str, album: str | None = None,
                  artist: str | None = None) -> dict | None:
        res = self.search3(title, song_count=50)
        return match_song(as_list(res.get("song")), title, album, artist)

    def scan_status(self) -> dict:
        return self._request("getScanStatus").get("scanStatus", {})

    def get_playlists(self) -> list:
        return as_list(self._request("getPlaylists").get("playlists", {}).get("playlist"))

    # write (used only by the prompted playlist-build step)
    def start_scan(self, full: bool = False) -> dict:
        params = {"fullScan": "true"} if full else {}
        return self._request("startScan", **params).get("scanStatus", {})

    def create_playlist(self, name: str, song_ids: list[str]) -> dict:
        # Navidrome preserves the submitted songId order.
        return self._request("createPlaylist", name=name, songId=list(song_ids)).get("playlist", {})


def client_from_env(dotenv: str | Path | None = None) -> Subsonic:
    if dotenv:
        load_dotenv(dotenv)
    url = os.environ.get("NAVIDROME_URL")
    user = os.environ.get("NAVIDROME_USER")
    pw = os.environ.get("NAVIDROME_PASS")
    if not (url and user and pw):
        raise SubsonicError("NAVIDROME_URL / NAVIDROME_USER / NAVIDROME_PASS not set (see .env.example)")
    return Subsonic(url, user, pw)


# --- cli (read-only smoke commands; run against the live homelab) ------------

def run_cli(args) -> int:
    repo_root = Path(__file__).resolve().parent.parent
    client = client_from_env(dotenv=repo_root / ".env")
    cmd = args.navidrome_cmd
    if cmd == "ping":
        client.ping()
        print("ok — Navidrome reachable and auth valid")
    elif cmd == "find-album":
        a = client.find_album(args.album, args.artist)
        if not a:
            print("album not found")
            return 1
        print(f"{a['name']} — {a.get('artist')}  id={a['id']}  songs={a.get('songCount')}")
    elif cmd == "album-songs":
        for s in songs_of_album(client.get_album(args.id)):
            print(f"  {int(s.get('track', 0)):>2}. {s.get('title')}  id={s['id']}")
    elif cmd == "scan-status":
        print(client.scan_status())
    return 0
